### PR TITLE
Tarball remote

### PR DESCRIPTION
**Pull request by kirillseva**
Support things like

```yml
-
    name: xgboost
    remote: tarball
    url: "http://dmlc.ml/drat/src/contrib/xgboost_0.6.4.6.tar.gz"
    version: 0.6.4.6
```

### CODE REVIEW

**Comment by abelcastilloavant** [comment]
If you add an example to the [README](https://github.com/robertzk/lockbox#example-lock-file), then it'll be easier for people to know about it and use it.

File: R/library.R [hunk 1]
```diff
@@ -272,6 +279,20 @@ download_package.github <- function(package, force) {
...
+download_package.tarball <- function(package, force) {
+  dest <- lockbox_package_download_path(package)
+  if (is.na(package$version)) {
+    package$version <- NULL
+  }
+
+  suppressWarnings(tryCatch(
+    download.file(url = package$url, destfile = lockbox_package_download_path(package)
+    , quiet = notTRUE(getOption('lockbox.verbose'))), error = function(e) e))
+
```
Comment (robertzk) [inline R/library.R:299]: Maybe double check here that the downloaded file is in fact a tarball.

**Review by robertzk — APPROVED** [review]
Looks good to me! Thank you.

File: R/library.R [hunk 2]
```diff
@@ -272,6 +279,20 @@ download_package.github <- function(package, force) {
   remote_download.github_remote(remote, dest, quiet = !isTRUE(getOption('lockbox.verbose')))
 }
 
+## download a raw tarball from a URL
+download_package.tarball <- function(package, force) {
```
Comment (abelcastilloavant) [inline R/library.R:290]: Why do we need both a downloading method and an installing method, when the installing method doesn't call out to the download method? (likely just that I don't know the layout of the lockbox package)

**Review by abelcastilloavant — APPROVED** [review]
Looks good to me. Thanks for updating the readme!

File: R/library.R (hunk 2 above)
Comment (robertzk) [reply R/library.R:290]: @abelcastilloavant This is a good point. However, note in the PR that the method dispatch for `get_remote_dependencies` for the tarball is the same as for CRAN, thus calling out to `download_package`.

File: R/library.R [hunk 3]
```diff
@@ -81,6 +81,13 @@ install_package.local <- function(locked_package, libPath, quiet) {
     , type = "source", quiet = quiet)
 }
 
+install_package.tarball <- function(locked_package, libPath, quiet) {
```
Comment (abelcastilloavant) [inline R/library.R:84]: Does it make sense to use the tarball that was downloaded when checking dependencies, in the same spirit as what we do with [CRAN packages](https://github.com/robertzk/lockbox/blob/master/R/library.R#L85-L89)?

Comment (robertzk) [reply R/library.R:84]: @kirillseva Can you address?

File: R/library.R (hunk 1 above)
Comment (kirillseva) [reply R/library.R:299]: how would I do that? I'm effectively downloading a binary from a URL I specified in the yaml file and hope that it's a tarball

File: R/library.R (hunk 2 above)
Comment (kirillseva) [reply R/library.R:290]: I don't understand why we download the packages (in order to get dependencies and all that) and then for the installation part we completely ignore the downloaded package and instead install from CRAN, github or URL directly, thereby redownloading the package. Seems like a little refactoring would help. I'm lost in the codebase myself, frankly

File: R/library.R (hunk 3 above)
Comment (kirillseva) [reply R/library.R:84]: will do, good catch! 👀

File: R/library.R [hunk 4]
```diff
@@ -82,10 +82,17 @@ install_package.local <- function(locked_package, libPath, quiet) {
 }
 
 install_package.tarball <- function(locked_package, libPath, quiet) {
-  stopifnot(is.element("url", names(locked_package)))
+  ## First, attempt to install the already downloaded tarball
```
Comment (robertzk) [inline R/library.R]: check if file exists

File: R/library.R [hunk 5]
```diff
@@ -82,10 +82,17 @@ install_package.local <- function(locked_package, libPath, quiet) {
...
+  try(utils::install.packages(locked_package$download_path, lib = libPath
+    , repos = NULL, type = "source", quiet = quiet))
```
Comment (robertzk) [inline R/library.R]: `silent=T`

**Review by robertzk — CHANGES_REQUESTED** [review]

File: R/library.R (hunk 5 above)
Comment (kirillseva) [reply R/library.R]: I do not see `silent` as one of the arguments to `utils::install.packages`. You mean somewhere else?

File: R/library.R (hunk 4 above)
Comment (kirillseva) [reply R/library.R]: This is handled by tryCatch already

**Review by robertzk — APPROVED** [review]

**Comment by robertzk** [comment]
@kirillseva LGTM if you tested from a fresh install.

**Pull request merged**